cdc_iterator.go

```diff
@@ -291,7 +414,37 @@ func newCdcEventHandler(
 
 func (h *cdcEventHandler) OnRow(e *canal.RowsEvent) error {
 	binlogName := h.canal.SyncedPosition().Name
-	rowEvent := rowEvent{e, binlogName}
+
+	rowEvent := rowEvent{
+		binlogName: binlogName,
+		table:      e.Table,
+		header:     e.Header,
+		action:     e.Action,
+		rows:       make([]replicationEventRow, 0, len(e.Rows)),
+	}
+
+	if e.Action == canal.UpdateAction && len(e.Rows)%2 != 0 {
+		return fmt.Errorf("even number of rows is expected in replication event")
+	}
+
+	for i := 0; i < len(e.Rows); i++ {
```

Comment:
May as well build the record here and send an opencdc record to the iterator, you have all the datas you need.
Splitting each action into its own function handler can help clean up the logic in the switch as well.